### PR TITLE
Add anonymization feature for private user data

### PR DESCRIPTION
**Pull request by codingjoe**
User privacy is important, not only to meet local regulations, but also to
protect your users and allow them to exercise their rights. However,
it's not always practical to delete users, especially if they have dependent
objects, that are relevant for statistical analysis.

Anonymization is a process of removing the user's personal data whilst keeping
related data intact. This is done by using the ``anonymize`` method.

### CODE REVIEW

File: mailauth/contrib/user/migrations/0005_emailuser_email_hash_alter_emailuser_email.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            ),
+        ),
+        migrations.RunSQL(
+            "UPDATE mailauth_user_emailuser SET email_hash = md5(email)",
+            "UPDATE mailauth_user_emailuser SET email_hash = NULL",
+        ),
+        migrations.AlterField(
+            model_name="emailuser",
+            name="email_hash",
+            field=models.TextField(db_index=True, default=0, max_length=16),
```
[the remark this answers is not in the document]
Comment (fidelitas) [reply mailauth/contrib/user/migrations/0005_emailuser_email_hash_alter_emailuser_email.py:40]: however, creating a user with hexdigest works well. I'm not sure what this migration operation actually is for then.

File: mailauth/contrib/user/models.py [hunk 2]
```diff
@@ -50,9 +53,24 @@ class AbstractEmailUser(AbstractUser):
     username = None
     password = None
 
-    email = CIEmailField(_("email address"), unique=True, db_index=True)
+    email = CIEmailField(
+        _("email address"), blank=True, null=True, unique=True, db_index=True
+    )
     """Unique and case insensitive to serve as a better username."""
 
+    email_hash = models.TextField(db_index=True, editable=False)
```
Comment (fidelitas) [inline mailauth/contrib/user/models.py:61]: The field will not be populated when updating the lib. Will it?

File: mailauth/contrib/user/models.py [hunk 3]
```diff
@@ -68,6 +86,19 @@ def has_usable_password(self):
     class Meta(AbstractUser.Meta):
         abstract = True
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._email = self.email
+
+    def save(self, *args, **kwargs):
+        if self.email and ((self._email != self.email) or not self.pk):
+            self.email_hash = hashlib.md5(self.email.lower().encode()).digest()
```
Comment (fidelitas) [inline mailauth/contrib/user/models.py:95]: Since you are testing for `hexdigest` in your tests, this bit is untested I guess.

**Review by fidelitas — COMMENTED** [review]
@codingjoe nice feature.
I think the real blockers hereare that you are using digest as well as hexdigest for a field you want to filter on and that you are not populating the new hash field. Which, I guess, is something you want.
I surely can't know all the implications this change will have on actual usage. Perhaps we should once install it @ voiio platform and test.

File: mailauth/contrib/user/models.py [hunk 4]
```diff
@@ -50,9 +53,24 @@ class AbstractEmailUser(AbstractUser):
...
+
+    GDPR may require us to erase the email address, yet we still want to be able to
+    retain non-personal information for statistical or other legal purposes. We may
+    also want to be able to authenticate users without ever storing the email address.
+
+    A hash, being non-reversible in nature, means that the personal information is
+    not stored and you may only processes personal information during runtime for
+    the explicit purpose of authenticating the user.
+    """
+
```
[the remark this answers is not in the document]
Comment (codingjoe) [reply mailauth/contrib/user/models.py:73]: Sphinx's apidocs support both, but the latter is preferable, as it doesn't conflict with class docstrings, which would also go before all class attributes.

File: mailauth/contrib/user/migrations/0005_emailuser_email_hash_alter_emailuser_email.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+                db_index=True,
+                max_length=254,
+                null=True,
+                unique=True,
+                verbose_name="email address",
+            ),
+        ),
+        migrations.RunSQL(
+            "UPDATE mailauth_user_emailuser SET email_hash = md5(email)",
+            "UPDATE mailauth_user_emailuser SET email_hash = NULL",
```
[the remark this answers is not in the document]
Comment (fidelitas) [reply mailauth/contrib/user/migrations/0005_emailuser_email_hash_alter_emailuser_email.py:39]: sorry, my bad. I forgot how `RunSQL` works. You can ignore this comment.

File: mailauth/contrib/user/models.py (hunk 3 above)
Comment (fidelitas) [reply mailauth/contrib/user/models.py:95]: changing my phrase here as well.
Your tests do test for `digest` but the code uses hexdigest in other places.

Comment (fidelitas) [reply mailauth/contrib/user/models.py:95]: ...and the md5 function on the migration also does hexdigest as a small comment showed.

File: mailauth/contrib/user/models.py (hunk 2 above)
Comment (codingjoe) [reply mailauth/contrib/user/models.py:78]: We have a migration for the model that we ship. If anyone built a subclass, I will include an upgrade guide in the release notes.

File: mailauth/contrib/user/models.py [hunk 6]
```diff
@@ -18,6 +20,7 @@ def _create_user(self, email, **extra_fields):
         """Create and save a user with the given email."""
         email = self.normalize_email(email)
         user = self.model(email=email, **extra_fields)
+        user.email_hash = hashlib.md5(email.lower().encode()).hexdigest()
```
[the remark this answers is not in the document]
Comment (codingjoe) [reply mailauth/contrib/user/models.py:40]: it's hexdigest, thx.

File: mailauth/contrib/user/migrations/0005_emailuser_email_hash_alter_emailuser_email.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (codingjoe) [reply mailauth/contrib/user/migrations/0005_emailuser_email_hash_alter_emailuser_email.py]: I added comments to the migration.

**Comment by codingjoe** [comment]
@fidelitas I decided to ditch the whole hash thing. Hash collisions are an unlikely problem, but there really is no need to retain a hash. So just anonymization now :)

**Review by fidelitas — APPROVED** [review]
I'm really impressed of the clean code (also the tests) that has been made here.
Very well done!

**Pull request merged**